from pymongo import MongoClient
from time import sleep

class Usuario:
    def __init__(self, nome, idade):
        self.nome = nome
        self.idade = idade

class GerenciadorUsuarios:
    def __init__(self):
        try:
            self.cliente = MongoClient('localhost', 27017)
            self.db = self.cliente['cadastro']
            self.collection = self.db['usuarios']
            print("CONEX√ÉO AO BANCO DE DADOS ESTABELECIDA COM SUCESSO!")
        except Exception as e:
            print(f"ERRO AO CONECTAR AO BANCO DE DADOS: {e}")

    def adicionar_usuario(self, nome, idade):
        try:
            usuario = {"nome": nome, "idade": idade}
            self.collection.insert_one(usuario)
            print("USU√ÅRIO ADICIONADO COM SUCESSO!")
        except Exception as e:
            print(f"ERRO AO ADICIONAR USU√ÅRIO: {e}")

    def listar_usuarios(self):
        try:
            usuarios = list(self.collection.find())

            if len(usuarios) > 0:
                print("=" * 100)
                print("LISTA DE USU√ÅRIOS:")
                print("-" * 100)
                for usuario in usuarios:
                    print("*" * 100)
                    print(f"NOME: {usuario['nome']}, IDADE: {usuario['idade']}")
                    print("*" * 100)
                print("=" * 100)
            else:
                print("NENHUM USU√ÅRIO CADASTRADO.")
        except Exception as e:
            print(f"ERRO AO LISTAR USU√ÅRIOS: {e}")

    def atualizar_usuario(self, nome_antigo, novo_nome, nova_idade):
        try:
            query = {"nome": nome_antigo}
            new_values = {"$set": {"nome": novo_nome, "idade": nova_idade}}
            self.collection.update_one(query, new_values)
            print("USU√ÅRIO ATUALIZADO COM SUCESSO!")
        except Exception as e:
            print(f"ERRO AO ATUALIZAR USU√ÅRIO: {e}")

    def excluir_usuario(self, nome):
        try:
            query = {"nome": nome}
            self.collection.delete_one(query)
            print("USU√ÅRIO EXCLU√çDO COM SUCESSO!")
        except Exception as e:
            print(f"ERRO AO EXCLUIR USU√ÅRIO: {e}")

def exibir_menu():
    print("\nMENU:")
    print("1. ADICIONAR USU√ÅRIO")
    print("2. LISTAR USU√ÅRIOS")
    print("3. ATUALIZAR USU√ÅRIO")
    print("4. EXCLUIR USU√ÅRIO")
    print("5. SAIR")

def main():
    gerenciador = GerenciadorUsuarios()

    while True:
        exibir_menu()
        opcao = input("ESCOLHA UMA OP√á√ÉO:\n>>>")

        if opcao == "1":
            nome = input("DIGITE O NOME:\n>>>")
            idade = input("DIGITE A IDADE:\n>>>")
            gerenciador.adicionar_usuario(nome, idade)
        elif opcao == "2":
            gerenciador.listar_usuarios()
        elif opcao == "3":
            nome_antigo = input("DIGITE O NOME A SER ATUALIZADO:\n>>>")
            novo_nome = input("DIGITE O NOVO NOME:\n>>>")
            nova_idade = input("DIGITE A NOVA IDADE:\n>>>")
            gerenciador.atualizar_usuario(nome_antigo, novo_nome, nova_idade)
        elif opcao == "4":
            nome = input("DIGITE O NOME DO USU√ÅRIO A SER EXCLU√çDO:\n>>>")
            gerenciador.excluir_usuario(nome)
        elif opcao == "5":
            print("üöÄSAINDO...")
            sleep(3)
            break
        else:
            print("OP√á√ÉO INV√ÅLIDA. TENTE NOVAMENTE!")

if __name__ == "__main__":
    main()
